addons/mail/models/mail_thread.py

```diff
@@ -2135,6 +2135,41 @@ def _message_create(self, values_list):
     # NOTIFICATION API
     # ------------------------------------------------------
 
+    @api.model
+    def discard_notifications(self, notification_type):
+        if notification_type in {'email', 'sms'}:
+            # TDE CHECK: delete pending SMS
+            author_id = self.env.user.partner_id.id
+            self._cr.execute("""
+                            SELECT notif.id, mes.id
+                            FROM mail_notification notif
+                            JOIN mail_message mes
+                                ON notif.mail_message_id = mes.id
+                            WHERE notif.notification_type = %(notification_type)s IS TRUE 
+                              AND notif.author_id = %(author_id)s
+                              AND notif.notification_status IN ('bounce', 'exception')
+                              AND mes.model = %(model_name)s
+                                """,
+                             {'model_name': self._name, 'author_id': author_id, 'notification_type': notification_type})
+            res = self._cr.fetchall()
+            notif_ids = [row[0] for row in res]
+            message_ids = list(set([row[1] for row in res]))
+            if notif_ids:
+                self.env['mail.notification'].browse(notif_ids).sudo().write({'notification_status': 'canceled'})
+            if message_ids:
+                self.env['mail.message'].browse(message_ids)._notify_message_notification_update()
+        elif notification_type == 'snail':
```

Comment:
This should be done by overrides. No need to have one for SMS as it basically does the same think as mail. But snailmail should have its own override, like in snailmail/models/mail_thread.py : 
```
def discard_notifications(self, notification_type):
    if notification_type != 'snail':
        return super().discard_notifications(notification_type)
    # do stuff specific to snailmail
```

Reply:
Done